wc-admin: Include package

psealock opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Add wc-admin as a dependency

### How to test the changes in this Pull Request:

1. Checkout this branch and run `composer update`
2. Check that wc-admin is present and working
3. Check version for conflicts

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enhancement - New WooCommerce Admin.

psealock committed:
wc-admin: Include package

psealock committed:
wc-admin: Include package

claudiosanches commented on bin/package-update.sh:
```diff
@@ -31,6 +31,8 @@ output 3 "Updating package textdomains..."
 find ./packages/woocommerce-blocks \( -iname '*.php' -o -iname '*.js' \) -exec sed -i.bak -e "s/'woo-gutenberg-products-block'/'woocommerce'/g" -e "s/\"woo-gutenberg-products-block\"/'woocommerce'/g" {} \;
 find ./packages/woocommerce-rest-api -iname '*.php' -exec sed -i.bak -e "s/, 'woocommerce-rest-api'/, 'woocommerce'/g" {} \;
+find ./packages/woocommerce-admin -iname '*.php' -exec sed -i.bak -e "s/, 'woocommerce-admin'/, 'woocommerce'/g" {} \;
+find ./packages/woocommerce-admin -iname '*.js' -exec sed -i.bak -e "s/, 'woocommerce-admin'/, 'woocommerce'/g" {} \;
```
We need a better script for it, it's easy now to override important data with this, since those may be slugs for another functionality.

timmyc replied:
@claudiosanches wouldn't this be true for the two lines above also? Do you have any ideas for a better approach?

claudiosanches replied:
@timmyc I have a solution for it already: https://github.com/woocommerce/woocommerce/pull/25536

psealock commented on tests/bootstrap.php:
```diff
@@ -82,6 +82,10 @@ public function install_wc() {
+		// Initialize the WC API extensions.
+		\Automattic\WooCommerce\Admin\Install::create_tables();
+		\Automattic\WooCommerce\Admin\Install::create_events();
+
```
@claudiosanches I needed to add this in order to bootstrap wc-admin tables for tests. It seems like probably not the best place, do you have a suggestion here?

claudiosanches replied:
Not sure what we can do.
@rodrigoprimo can you help us?

rodrigoprimo replied:
IMO we can leave it as is for now. I agree that is not great but I don't think we currently have an alternative and I don't think it is worth creating something now.

psealock committed:
phpcs errors

psealock committed:
phpcs errors

psealock committed:
cleanup

psealock committed:
cleanup

jeffstieler committed:
Alias a branch of wc-admin instead.

jeffstieler committed:
Alias a branch of wc-admin instead.

timmyc commented:
While testing this I've gotten the following when attempting to DELETE the woo plugin. I have reproduced this 2x now, but unfortunately there is nothing showing in the PHP error log:

<img width="1520" alt="image" src="https://user-images.githubusercontent.com/22080/72195443-2c1fa000-33c7-11ea-8587-89f284614743.png">

jeffstieler committed:
Load the WooCommerce Admin install class during uninstallation.

jeffstieler committed:
Load the WooCommerce Admin install class during uninstallation.

jeffstieler commented:
@timmyc - wc-admin tables should now get cleaned up, see: 01ba05a

jeffstieler committed:
Fix wc_admin_daily event schedule persistence.

Avoid an option update race condition.

jeffstieler committed:
Fix wc_admin_daily event schedule persistence.

Avoid an option update race condition.

timmyc commented:
Looks like we should get a rebase here again :)

jeffstieler commented:
@timmyc rebased in 4b1ccc4

jeffstieler committed:
Fix unit tests.

Only require autoloader in uninstall if WooCommerce Admin class hasn't been loaded.

jeffstieler committed:
Fix unit tests.

Only require autoloader in uninstall if WooCommerce Admin class hasn't been loaded.

jeffstieler committed:
Fix WP global override.

jeffstieler committed:
Fix WP global override.

jeffstieler committed:
Update woocommerce-admin dependency.

jeffstieler commented:
E2E tests are passing for me locally, but seem to be failing in CI..

peterfabian commented:
I tested it out and the integration seems to be working, will test some more plus asked team to have a look individually as well.

Is this one ready for review or is there something missing? Is there a plan to include the off switch or not?

jeffstieler commented:
> Is this one ready for review or is there something missing?

@peterfabian yes, this is ready for a review, but not a merge (yet). We'd want to verify our approach here with the changes to core first, then we can update WooCommerce Admin to include the pending changes to support this branch, and then target a release number rather than a Git branch in the composer file here.

> Is there a plan to include the off switch or not?

If you're referring to [this filter](https://github.com/woocommerce/woocommerce-admin/blob/7051557a69f3514a26d1d02acb52e87343112faa/src/FeaturePlugin.php#L62), that functionality is still intact.

peterfabian commented on composer.json:
```diff
@@ -12,8 +12,15 @@
     "woocommerce/woocommerce-blocks": "2.5.10",
-    "woocommerce/woocommerce-rest-api": "1.0.6"
+    "woocommerce/woocommerce-rest-api": "1.0.6",
+    "woocommerce/woocommerce-admin": "dev-try/downgrade-autoloader as 0.24.0"
```
Is this the correct version we plan to include?

jeffstieler replied:
We need to update the autoloader version in `master` first.

timmyc replied:
@jeffstieler what version do we need to update it to? Background here p90Yrv-1wu-p2

jeffstieler replied:
@timmyc we don't yet have a PR to update the version in WC-Admin, but it needs to change to `^1.2.0` to match core WC and the blocks repo.

timmyc replied:
So just a quick check, it looks like `master` [is set to `^1.2.0`](https://github.com/woocommerce/woocommerce/blob/master/composer.json#L11) so thinking this requirement has been satisfied.

jeffstieler replied:
It needs to be updated to match in the WC-Admin repo as well.

timmyc replied:
https://github.com/woocommerce/woocommerce-admin/pull/3667/

jeffstieler replied:
Updated in 88ec069.

timmyc replied:
v0.25.0 has been published, and jetpack-auotloader in wc-admin is the same version as here in Woo core. ( cc @peterfabian  )

peterfabian replied:
Thanks for the update! The current inclusion uses the source package that is getting cloned from GitHub, not the dist package. Can we switch to dist/zip package, similarly to how Blocks is doing it? Otherwise wc-admin is 47MB when cloned to `packages/`.

timmyc replied:
Correct I believe we were pointed at GitHub to resolve the issue we were having with the mis-match of the jetpack autoloader. So we should be able to point at packagist ( as @claudiosanches mentions below )

claudiosanches replied:
Solved!

jeffstieler committed:
Use release version of WooCommerce Admin.

peterfabian committed:
Merge branch 'master' into try/add-wc-admin

claudiosanches commented on composer.json:
```diff
@@ -14,8 +14,15 @@
+    "woocommerce/woocommerce-rest-api": "1.0.7",
+    "woocommerce/woocommerce-admin": "0.25.0"
   },
+  "repositories": [
```
We can remove this now, since WooCommerce Admin is published in https://packagist.org/packages/woocommerce/woocommerce-admin

claudiosanches committed:
Merge branch 'try/add-wc-admin' of github.com:woocommerce/woocommerce into try/add-wc-admin

claudiosanches committed:
Removed old repository entry from composer.json

claudiosanches committed:
Fixed coding standards

claudiosanches committed:
Merge branch 'master' into try/add-wc-admin

claudiosanches committed:
Merge branch 'master' into try/add-wc-admin

claudiosanches commented on uninstall.php:
```diff
@@ -28,6 +28,24 @@
+		// Hook in WooCommerce Admin installation code.
+		// @todo an alternate way to do this is allow FeaturePlugin::has_satisfied_dependencies()
+		// to work during uninstall - (no WooCommerce instance).
+		\Automattic\WooCommerce\Admin\Install::init();
```
@woocommerce/isotope I think it suppose to be `drop_tables()` instead of `init()`.

rodrigoprimo commented on tests/bootstrap.php:
```diff
@@ -82,11 +85,15 @@ public function install_wc() {
 		WC_Install::install();
 
-		// Reload capabilities after install, see https://core.trac.wordpress.org/ticket/28374
+		// Initialize the WC API extensions.
```
This comment updated (or removed as to me it seems straigthforward to understand what happening in the two lines below just by reading the class and method names).

rodrigoprimo commented on tests/bootstrap.php:
```diff
@@ -82,11 +85,15 @@ public function install_wc() {
-		// Reload capabilities after install, see https://core.trac.wordpress.org/ticket/28374
+		// Initialize the WC API extensions.
+		\Automattic\WooCommerce\Admin\Install::create_tables();
+		\Automattic\WooCommerce\Admin\Install::create_events();
```
Why do we need to initialize the events? Maybe just create the tables is enough for unit test purposes?

rodrigoprimo commented on uninstall.php:
```diff
@@ -28,6 +28,24 @@
+	/**
+	 * Load core packages autoloader.
+	 */
+	if ( version_compare( PHP_VERSION, '5.6.0', '>=' ) ) {
```
Maybe this check is not needed anymore now that WooCommerce core requires PHP >= 7.0?

rodrigoprimo commented on uninstall.php:
```diff
@@ -28,6 +28,24 @@
+	 * Load core packages autoloader.
+	 */
+	if ( version_compare( PHP_VERSION, '5.6.0', '>=' ) ) {
+		if ( ! class_exists( 'Automattic\WooCommerce\Admin\Install' ) ) {
```
Why is this if statement needed? I didn't check but I believe that before calling uninstall.php, WP will load the plugin, so it should be safe to assume that the autoloader has been initialized. Or am I missing something? If we are worried that for some reason it is not possible to initialize the autoloader and then call `\Automattic\WooCommerce\Admin\Install::init()` (or more likely `drop_tables()` as Claudio mentioned below), maybe it is better to check if the method exist before calling it? My concern is that if we return here, the code that handles the removal of the rest of WooCommerce will not run.

rodrigoprimo replied:
Ignore the part where I said that WP loads the plugin before calling uninstall.php. I just checked and that is not the case.

rodrigoprimo commented on uninstall.php:
```diff
@@ -72,21 +90,21 @@
 	if ( version_compare( $wp_version, '4.2', '>=' ) ) {
 		// Delete term taxonomies.
-		foreach ( array( 'product_cat', 'product_tag', 'product_shipping_class', 'product_type' ) as $taxonomy ) {
+		foreach ( array( 'product_cat', 'product_tag', 'product_shipping_class', 'product_type' ) as $_taxonomy ) {
```
Good catch, thanks!

peterfabian committed:
Update the uninstall routine to fix review issues.

- PHP version check should not be needed anymore, as WC core requires PHP 7.0+
- The uninstallation/cleanup is not aborted if WC Admin can't be loaded
- Replaced init() call by drop_tables() call, as that's what probably needs to happen

peterfabian commented:
I've tried to address the review items raised here in https://github.com/woocommerce/woocommerce/pull/25011/commits/7c81915d1cf26a62ac85d75eefb394b307e0f9f2. Time for The Merge™️.

peterfabian reviewed (approved):
I think we're ready.
